src/service/pom.rs

```diff
@@ -34,79 +37,145 @@ impl crate::server::MyService {
+                        protocol_name
                     )));
                 }
             }
-        }
-        let toml = match std::fs::read_to_string(&path).unwrap().parse::<Value>() {
-            Ok(toml) => toml,
-            Err(err) => return Err(Status::internal(err.to_string())),
-        };
-        if self.core_uri.is_none() {
-            return Err(Status::internal("core_uri not found."));
-        }
-        let core_addr = self.core_uri.as_ref().unwrap();
-        if toml.get("package").is_none() || toml["package"].get("entrypoint").is_none() {
-            return Err(Status::not_found("entrypoint not found."));
-        }
-        let entrypoint = toml["package"]["entrypoint"].as_str();
-        if entrypoint.is_none() {
-            return Err(Status::not_found("entrypoint not found."));
-        }
-        let entrypoint = entrypoint.unwrap();
-        let user_jwt = self
-            ._host_storage_read(&format!("users:{}:user_jwt", request.get_ref().user_id))
+            let toml_path = protocol_home.join("colink.toml");
+            if std::fs::metadata(&toml_path).is_err() {
+                match fetch_protocol_from_inventory(protocol_name, &colink_home).await {
+                    Ok(_) => {}
+                    Err(err) => {
+                        return Err(Status::not_found(&format!(
+                            "protocol {} not found: {}",
+                            protocol_name, err
+                        )));
+                    }
+                }
+            }
+            let toml = match std::fs::read_to_string(&toml_path)
+                .unwrap()
+                .parse::<Value>()
+            {
+                Ok(toml) => toml,
+                Err(err) => return Err(Status::internal(err.to_string())),
+            };
+            if self.core_uri.is_none() {
+                return Err(Status::internal("core_uri not found."));
+            }
+            let core_addr = self.core_uri.as_ref().unwrap();
+            if toml.get("package").is_none() || toml["package"].get("entrypoint").is_none() {
+                return Err(Status::not_found("entrypoint not found."));
+            }
+            let entrypoint = toml["package"]["entrypoint"].as_str();
+            if entrypoint.is_none() {
+                return Err(Status::not_found("entrypoint not found."));
+            }
+            let entrypoint = entrypoint.unwrap();
+            if toml["package"].get("install_script").is_some() {
+                let install_script = toml["package"]["install_script"].as_str();
+                if install_script.is_none() {
+                    return Err(Status::internal("invalid install_script"));
+                }
+                let install_script = install_script.unwrap();
+                let init_file = protocol_home.join(".install_timestamp");
+                if std::fs::metadata(&init_file).is_err() {
+                    match Command::new("bash")
```

Comment:
shall we consider making a wrapping of this? (seems to appear >= 3 times in the code base)